[Keane] iP - Pilcrow task manager

okyntary opened the pull request:
# Pilcrow - the future of task management

*Disclaimer: the claims below are made in jest (obviously). Do no take them seriously.*

> Pilcrow is the future of task management. It is my magnum opus, my Mona Lisa, my piece de resistance. It combines function and form in such an elegant way that one who has laid eyes or ears upon it can scarcely fathom any possible state for it, besides sheer sublimity and ineffable perfection. In short, it is nothing but the culmination and precipitation of my life's work, the single point towards which all my efforts since birth have been directed. It is magnificence in its raw unfiltered form; we are all simlarly graced by its presence. -  Me

🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩


Welcome to Pilcrow, the most advanced task-keeping software you'll find anywhere!

Here's a list of features that are already available in the current version of Pilcrow - you can:
1. Add tasks!
1. List your existing tasks out!
1. Mark tasks as done!
1. Delete tasks!
1. And much, much, much more!

You can even tinker around with it to improve it to your own liking too! Here's a snippet of Java code that I used:

~~~java
    public static void main(String[] args) {
        Pilcrow pilcrow = new Pilcrow();
        pilcrow.run();
    }
~~~

Wow! Aren't you impressed?

Here's a live demo of how Pilcrow works: [Pilcrow Demo](https://www.theraleighregister.com/wwwyoutubecomamazingpilcrowdemo.html)

Isn't it great? Don't say I didn't warn you!

How did you feel after watching the demo?
- [x] Awestruck
- [x] Dumbstruck
- [x] I have glimpsed a fraction of the divine, a sliver of transcendence, a glimpse of God. I shall now download Pilcrow and use it for the remainder of my days.

Still not convinced? Here's some reasons why Pilcrow is so great:
- It's free.
- It's cheap. In fact, it's so cheap it'll cost you nothing. It's free, really!
- You can use it on a computer. Computers are cool, right? They're *in* now, aren't they?
- You operate it by typing in words. Words allow you to say things and communicate information, otherwise we wouldn't be able to tell each other anything. Since words are important, Pilcrow is important too!

This is just a fraction of what Pilcrow can do for you; there' lots more to come! Don't take my word for it, just try it!

okyntary committed:
Rename Duke.java to Pilcrow.java

okyntary committed:
Pilcrow: Add add and list functionalities

okyntary committed:
Pilcrow: Add task functionality

okyntary committed:
Pilcrow: add different types of tasks, refactor code

okyntary committed:
Modify and update text-ui-test files to test Pilcrow.java

okyntary committed:
Pilcrow: Add exception handling for faulty inputs into Pilcrow

okyntary committed:
Pilcrow: add delete functionality

okyntary committed:
Overhaul Pilcrow to include more abstraction and OOP concepts

okyntary committed:
Fix minor issues in Storage.java and Pilcrow.java

okyntary committed:
Add storage of tasks onto the hard drive

okyntary committed:
Add LocalDate functionality to Deadline events

okyntary committed:
Merge branch 'branch-Level-7'

okyntary committed:
Merge branch-Level-8 into master

okyntary committed:
Organize all classes into the Pilcrow package

okyntary committed:
Add basic JUnit tests

okyntary committed:
Create .jar file for Pilcrow

okyntary committed:
Pilcrow: Add search functionality, allowing user to search tasks for keyword

okyntary committed:
Reformat code to follow se-edu Java coding standard

okyntary committed:
Add JavaDocs to all files in Pilcrow package

okyntary committed:
Merge branch 'branch-A-JavaDoc'

okyntary committed:
Merge branch 'branch-A-CodingStandard'

okyntary committed:
Merge tag 'Level-9'

okyntary committed:
Use GFMD in the PR description

jovyntls commented on src/main/java/pilcrow/Parser.java:
```diff
@@ -0,0 +1,60 @@
+     */
+    public int getIndex() {
+        int index;
+        if (restOfCommand.equals("")) {
```
```suggestion
        if (this.restOfCommand.equals("")) {
```

Since `restOfCommand` is an instance variable, would it be clearer to use `this`?
Applies to line 50 too.

jovyntls commented on src/main/java/pilcrow/Parser.java:
```diff
@@ -0,0 +1,60 @@
+     * Used for commands that expect an integer to follow, e.g. done, delete.
+     * @return Integer following initial word.
+     */
+    public int getIndex() {
```
```suggestion
    public int getIndex() throws InvalidInputException {
```
Since `InvalidInputException` is thrown in line 46, would it be better to `throw` it in the method description?

jovyntls commented on src/main/java/pilcrow/Parser.java:
```diff
@@ -0,0 +1,60 @@
+            } else {
+                index = Integer.parseInt(restOfCommand);
+            }
+        } catch (NumberFormatException exception) {
+            throw new InvalidInputException("Must enter integer following command.");
+        }
```
I like the detailed use of Exceptions. It makes it very clear and easier for debugging!

jovyntls commented on src/main/java/pilcrow/Pilcrow.java:
```diff
@@ -0,0 +1,91 @@
+                        this.ui, this.storage, this.taskList, parser);
+
+                isExit = (parser.getCommandWord().equals("bye"));
+                if (isExit) {
+                }
```
Is there supposed to be some logic in this `if` block?

auhc99 replied:
You should remove this block if there is no use for it.

okyntary replied:
Good idea, I'll do that.

jovyntls commented on src/main/java/pilcrow/Storage.java:
```diff
@@ -0,0 +1,70 @@
+            FileWriter pilcrowFileWriter = new FileWriter(this.filePath);
+            pilcrowFileWriter.write(storedTasks);
+            pilcrowFileWriter.close();
+        }  catch (IOException exception) {
+
+        }
```
Is this block left intentionally empty?

jovyntls reviewed (commented):
Looks good 👍

okyntary committed:
Implement checkstyle in Gradle

auhc99 commented on README.md:
```diff
@@ -1,24 +1,45 @@
+
+🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩🤩
+
+
+Welcome to Pilcrow, the most advanced task-keeping software you'll find anywhere!
+
+Here's a list of features that are already available in the current version of Pilcrow - you can:
+1. Add tasks!
+1. List your existing tasks out!
+1. Mark tasks as done!
+1. Delete tasks!
+1. And much, much, much more!
+
+You can even tinker around with it to improve it to your own liking too! Here's a snippet of Java code that I used:
+
+~~~java
+    public static void main(String[] args) {
+        Pilcrow pilcrow = new Pilcrow();
+        pilcrow.run();
+    }
+~~~
+
+Wow! Aren't you impressed?
+
+Here's a live demo of how Pilcrow works: [Pilcrow Demo](https://www.theraleighregister.com/wwwyoutubecomamazingpilcrowdemo.html)
+
+Isn't it great? Don't say I didn't warn you!
+
+How did you feel after watching the demo?
+- [x] Awestruck
+- [x] Dumbstruck
+- [x] I have glimpsed a fraction of the divine, a sliver of transcendence, a glimpse of God. I shall now download Pilcrow and use it for the remainder of my days.
+
+Still not convinced? Here's some reasons why Pilcrow is so great:
+- It's free.
+- It's cheap. In fact, it's so cheap it'll cost you nothing. It's free, really!
+- You can use it on a computer. Computers are cool, right? They're *in* now, aren't they?
+- You operate it by typing in words. Words allow you to say things and communicate information, otherwise we wouldn't be able to tell each other anything. Since words are important, Pilcrow is important too!
+
+This is just a fraction of what Pilcrow can do for you; there' lots more to come! Don't take my word for it, just try it!
```
I love your README!

auhc99 commented on src/main/java/pilcrow/Task.java:
```diff
@@ -0,0 +1,208 @@
+     * @param isDone Done status of the Task to be created.
+     * @return Task with specified fields.
+     */
+    public static Task createTask(String taskType, String taskName, Boolean isDone) {
```
You can consider declaring an enum for the TaskType.

auhc99 commented on src/main/java/pilcrow/TaskList.java:
```diff
@@ -0,0 +1,96 @@
+        return this.taskList.get(index);
+    }
+
+    public TaskList filteredTaskList(String string) {
```
You should rename your method such that the first word in the name is a verb.
For this case, you may consider renaming it to getFilteredTaskList.

okyntary replied:
Good idea, changed.

auhc99 reviewed (commented):
Overall great job with adhering to the Java coding standards!
Just a few nitpicks here and there from me that you can consider changing.

okyntary committed:
Implement Gradle support into Pilcrow

okyntary committed:
Add a basic GUI to Pilcrow

okyntary committed:
Fix small coding standard issues

okyntary committed:
Improve code quality

Code quality has not been maintained after additional functionality has
been added to Pilcrow.

Let's improve method and class comments, and also look for places where
we can refactor the code.

okyntary committed:
Add several assert statements

okyntary committed:
Practice writing commit messages.

I have not had to write full commit messages before.

Let's write a full commit message to practice this, before I move on to
bigger projects which require more documentation.

okyntary committed:
Merge branch 'master' into branch-A-CodeQuality

okyntary committed:
Merge pull request #1 from okyntary/branch-A-Assertions

Add several assert statements

okyntary committed:
Merge pull request #2 from okyntary/branch-A-CodeQuality

Improve code quality

xlunarin commented on src/main/java/pilcrow/Task.java:
```diff
@@ -0,0 +1,232 @@
+        return this.taskName.contains(string);
+    }
+
+    public void setIsDone(Boolean isDone) {
```
Perhaps this method name could be `setDone` instead of `setIsDon` as it would read more naturally!

xlunarin commented on src/main/java/pilcrow/Task.java:
```diff
@@ -0,0 +1,232 @@
+        return this.taskName;
+    }
+
+    public Boolean getIsDone() {
```
This could also be `getDone` instead of `getIsDone` to improve readability 😃

okyntary committed:
Add sort functionality for task lists in Pilcrow

xlunarin commented on src/main/java/pilcrow/Task.java:
```diff
@@ -0,0 +1,232 @@
+    private static class Deadline extends Task {
+        private LocalDate deadline;
+
+        private Deadline(String taskName, Boolean isDone) {
```
Maybe you should specify the InvalidInputExcetion thrown in the constructor signature? Since it is not being caught and handled inside the constructor itself.

```java
private Deadline(String taskName, Boolean isDone) throws InvalidInputException {
    // ...
}
```

xlunarin reviewed (commented):
Nice code overall!! Very neat, clean, and easy to read. 😁

okyntary committed:
Improve GUI and add Ui.png to /docs

okyntary committed:
Add User Guide in /docs/README.md

okyntary committed:
Change Ui.PNG to Ui.png

okyntary committed:
Add Ui.png

okyntary committed:
Delete Ui.PNG
